iothub/device/src/Pipeline/RetryDelegatingHandler.cs

```diff
@@ -435,6 +486,70 @@ public override async Task OpenAsync(CancellationToken cancellationToken)
             }
         }
 
+        private void StartLoopAsync(DateTime refreshOn, CancellationToken cancellationToken)
+        {
+            if (Logging.IsEnabled)
+                Logging.Enter(this, nameof(StartLoopAsync));
+
+            if (_loopCancellationTokenSource == null
+                || _refreshLoop == null)
+            {
+                (this as IAmqpAuthenticationRefresher)?.StopLoop();
+            }
+
+            _loopCancellationTokenSource = new CancellationTokenSource();
+            _refreshLoop = RefreshLoopAsync(refreshOn, cancellationToken);
+
+            if (Logging.IsEnabled)
+                Logging.Exit(this, nameof(StartLoopAsync));
+        }
+
+        private async Task RefreshLoopAsync(DateTime refreshesOn, CancellationToken cancellationToken)
+        {
+            TimeSpan waitTime = refreshesOn - DateTime.UtcNow;
+
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                if (Logging.IsEnabled)
+                    Logging.Info(this, refreshesOn, $"Before {nameof(RefreshLoopAsync)} with wait time {waitTime}.");
+
+                if (waitTime > TimeSpan.Zero)
+                {
+                    await Task.Delay(waitTime, cancellationToken).ConfigureAwait(false);
+                }
+
+                if (!cancellationToken.IsCancellationRequested)
+                {
+                    refreshesOn = await RefreshTokenAsync(cancellationToken).ConfigureAwait(false);
+                }
+
+                if (Logging.IsEnabled)
+                    Logging.Info(this, refreshesOn, $"After {nameof(RefreshLoopAsync)}");
+
+                waitTime = refreshesOn - DateTime.UtcNow;
+            }
+        }
+
+        private async void StopLoopAsync()
```

Comment:
Why `void` return, not Task?

Reply:
In StartLoop , I create a Task called _refreshLoop. I am not sure how returning a Task would be useful.

Reply:
<https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#async-void>